connstring.go

```diff
@@ -0,0 +1,400 @@
+	q.Values["onInit"] = P.onInitStmts
+	q.Add("configDir", P.ConfigDir)
+	q.Add("libDir", P.LibDir)
+	return quoteRunes(P.Username, "/@") + "/" + quoteRunes(password, "@") + "@" + P.CommonParams.DSN + "\n" + q.String()
+}
+
+// ParseConnString parses the given connection string into a struct.
+func ParseConnString(connString string) (ConnectionParams, error) {
+	P := ConnectionParams{
+		StandaloneConnection: DefaultStandaloneConnection,
+		CommonParams: CommonParams{
+			Timezone: time.Local,
+		},
+		ConnParams: ConnParams{
+			ConnClass: DefaultConnectionClass,
+		},
+		PoolParams: PoolParams{
+			MinSessions:      DefaultPoolMinSessions,
+			MaxSessions:      DefaultPoolMaxSessions,
+			SessionIncrement: DefaultPoolIncrement,
+			MaxLifeTime:      DefaultMaxLifeTime,
+			WaitTimeout:      DefaultWaitTimeout,
+			SessionTimeout:   DefaultSessionTimeout,
+		},
+	}
+
+	var paramsString string
+	connString = strings.TrimSpace(connString)
+	if i := strings.IndexByte(connString, '\n'); i >= 0 {
+		connString, paramsString = strings.TrimSpace(connString[:i]), strings.TrimSpace(connString[i+1:])
+	}
+	origConnString := connString
+	var q url.Values
+
+	if !strings.HasPrefix(connString, "oracle://") {
+		up, connString := splitQuoted(connString, '@')
+		if connString == "" {
+			return P, errors.New("no '@' in connection string")
+		}
+		P.Username, P.Password.secret = splitQuoted(up, '/')
+		P.Username, P.Password.secret = unquote(P.Username), unquote(P.Password.secret)
+
+		uSid := strings.ToUpper(connString)
+		//fmt.Printf("connString=%q SID=%q\n", connString, uSid)
+		if strings.Contains(uSid, " AS ") {
+			if P.IsSysDBA = strings.HasSuffix(uSid, " AS SYSDBA"); P.IsSysDBA {
+				connString = connString[:len(connString)-10]
+			} else if P.IsSysOper = strings.HasSuffix(uSid, " AS SYSOPER"); P.IsSysOper {
+				connString = connString[:len(connString)-11]
+			} else if P.IsSysASM = strings.HasSuffix(uSid, " AS SYSASM"); P.IsSysASM {
+				connString = connString[:len(connString)-10]
+			}
+		}
+		P.DSN = connString
+		/*
+			P.comb()
+			if Log != nil {
+				Log("msg", "ParseConnString", "connString", origConnString, "commonParams", P.CommonParams, "connParams", P.ConnParams, "poolParams", P.PoolParams)
+			}
+			return P, nil
+		*/
+		q = make(url.Values, 32)
+	} else {
+		u, err := url.Parse(connString)
+		if err != nil {
+			return P, errors.Errorf("%s: %w", connString, err)
+		}
+		if usr := u.User; usr != nil {
+			P.Username = usr.Username()
+			P.Password.secret, _ = usr.Password()
+		}
+		P.DSN = u.Hostname()
+		// IPv6 literal address brackets are removed by u.Hostname,
+		// so we have to put them back
+		if strings.HasPrefix(u.Host, "[") && !strings.Contains(P.DSN[1:], "]") {
+			P.DSN = "[" + P.DSN + "]"
+		}
+		if u.Port() != "" {
+			P.DSN += ":" + u.Port()
+		}
+		if u.Path != "" && u.Path != "/" {
+			P.DSN += u.Path
+		}
+		q = u.Query()
+	}
+	if paramsString != "" {
+		d := logfmt.NewDecoder(strings.NewReader(paramsString))
```

Comment:
logfmt does it support lists like ShardingKey ? may be we dont need it in this format?

Reply:
comment can be can be ignored, i believed lists is complex to handle as arguments. connector/openDb interface should  be able to acheive this.

Reply:
What kind of lists are you thinking about?
I don't know of any parameter that accepts a list, at the moment.

Reply:
He is referring to sharding keys. You can pass multiple values although it is more common to use just one.

Reply:
Ah, ok. Now it accepts only one, but it's a very simple fix (see handling of onInit).